### PR TITLE
fix(agent-loop): retry when model emits reasoning-only response

### LINKED ISSUE #705: bug: Mistral Vibe Stalls After "Thought" Stage—No Further Output or Steps
### Component

CLI

### Summary

Mistral Vibe frequently stops responding after displaying the Thought stage. The process initiates (e.g., showing a command or partial output), but abruptly halts without completing the task or providing further steps

Image

### Reproduction steps

- Input a command or query (e.g., find /Users/... or complex logic).
- Mistral Vibe displays the Thought stage with partial output.
- Process stops unexpectedly—no error message, no further output, and no completion.

### Versions / environment

vibe 2.10.1

### Logs & screenshots

```shell

```

### PR DESCRIPTION
**Pull request by Srujanrana07**
<html>
<body>
<html><head></head><body><h2>Fixes #705 — Agent stops responding after the Thought stage</h2>
<h3>What was happening</h3>
<p>When using a reasoning model (e.g. Magistral), the agent would show the <code>Thought ►</code> block in the UI and then silently stop — no output, no error, no further steps. The user is left staring at an empty output panel with no indication of what went wrong.</p>
<h3>Root Cause</h3>
<p>The model was emitting a <strong>thinking block only</strong> — <code>reasoning_content</code> was present, but <code>content</code> was empty and <code>tool_calls</code> was <code>None</code>. This is normal behaviour for reasoning models: they emit a thinking block as a standalone turn, expecting the caller to feed it back as context for the actual response.</p>
<p>Vibe was not handling this case. Here is the exact failure path:</p>
<ol>
<li><code>_stream_assistant_events()</code> correctly yields <code>ReasoningEvent</code>s — the Thought block appears in the UI</li>
<li><code>_chat_streaming()</code> appends the aggregated assistant message to history</li>
<li>Back in <code>_perform_llm_turn()</code>: <code>parse_message()</code> finds no content and no tool calls → returns early</li>
<li>Back in <code>_conversation_loop()</code>: last message has <code>role == assistant</code> → <code>should_break_loop = True</code> → <strong>silent exit</strong></li>
</ol>
<p>The thinking context was never fed back to the model, so it never got a chance to produce a real reply.</p>
<h3>What was changed</h3>
<p><strong><code>vibe/core/agent_loop.py</code></strong></p>
<p>Added a <code>_is_reasoning_only_message()</code> static helper that identifies the stall condition — an assistant message with <code>reasoning_content</code> but no <code>content</code> and no <code>tool_calls</code>.</p>
<p>Wrapped <code>_perform_llm_turn()</code> in a retry loop (capped at <code>_MAX_REASONING_ONLY_RETRIES = 3</code>):</p>
<ul>
<li>When a reasoning-only response is detected, the model is re-invoked. Since the thinking block is already appended to message history, the model gets it as context and can produce a real reply on the next turn.</li>
<li>If the retry budget is exhausted (pathological case), a clear <code></code> warning is surfaced to the user instead of hanging indefinitely.</li>
<li>Normal responses (content present, or tool calls present) hit the <code>break</code> immediately — zero overhead on the happy path.</li>
</ul>
<p><strong><code>tests/test_reasoning_only_hang.py</code></strong></p>
<p>10 new tests covering:</p>
<ul>
<li><code>_is_reasoning_only_message()</code> helper correctness across all edge cases</li>
<li>Core retry recovery in streaming mode (the main regression test)</li>
<li>Core retry recovery in non-streaming mode</li>
<li>Normal responses are completely unaffected — no extra LLM calls</li>
<li>Exhausting the retry budget produces a visible <code></code> error event</li>
</ul>
<h3>Before / After</h3>

Scenario | Before | After
-- | -- | --
Model emits reasoning-only on turn 1, content on turn 2 | Stops silently after turn 1 | Retries → surfaces content to user 
Model emits reasoning + content together | Works | Works, no change 
Model emits content only | Works | Works, no change 
Model keeps emitting reasoning-only (pathological) | Hangs forever | Shows  warning after 3 retries

### CODE REVIEW

**Comment by Srujanrana07** [comment]
Hey, just wanted to bump this — the PR has been sitting for a while and would love to get some eyes on it. It fixes a pretty noticeable issue where the agent silently stops after the Thought stage which affects anyone using a reasoning model. Happy to make any changes if needed!

File: vibe/core/agent_loop.py [hunk 1]
```diff
@@ -0,0 +1,1873 @@
...
+                # Exceeded retry budget — surface a visible error and bail out.
+                yield AssistantEvent(
+                    content=(
+                        "⚠️  The model completed its reasoning phase but did not "
+                        "produce a response after "
+                        f"{self._MAX_REASONING_ONLY_RETRIES} retries.  "
+                        "Try rephrasing your request or switching to a different model."
+                    )
+                )
+                return
```
Comment (Debasish-87) [inline vibe/core/agent_loop.py:1025]: Quick question on this `return` — it skips straight out of `_perform_llm_turn(),` which means the cleanup that normally runs at the end of this method (parse_message, `resolve_tool_calls,` the conditional `_setup_middleware(),` and await `self._reset_session())` never executes on this particular path. Before this change every call would reach that cleanup, so I wanted to flag it in case it's unintentional.

I don't know what `_reset_session()` does internally, so this may well be fine — but if it handles anything stateful (resetting connections, clearing cached state, etc.), skipping it here could leave the session in a slightly inconsistent state after the warning is shown. Happy to be corrected if this is deliberate, just wanted to make sure it wasn't missed.
